File: go/internal/httpresponse/response.go
// Package httpresponse builds the limited HTTP/1.1 responses used in this
// course without relying on net/http.
package httpresponse

import (
	"io"
	"strconv"
	"strings"
)

// Header is one HTTP response header field.
type Header struct {
	Name  string
	Value string
}

// Response is a byte-oriented HTTP/1.1 response description.
type Response struct {
	statusCode int
	reason     string
	headers    []Header
	body       []byte
}

// New creates a response. Bytes serializes it with a generated Content-Length.
func New(statusCode int, reason string, headers []Header, body []byte) Response {
	return Response{
		statusCode: statusCode,
		reason:     reason,
		headers:    append([]Header(nil), headers...),
		body:       append([]byte(nil), body...),
	}
}

// BasicOK returns the course's fixed text response.
func BasicOK() Response {
	return New(200, "OK", []Header{
		{Name: "Content-Type", Value: "text/plain"},
		{Name: "Connection", Value: "close"},
	}, []byte("Hello, World!"))
}

// Bytes serializes status line, headers, a generated Content-Length, and body.
func (r Response) Bytes() []byte {
	bytes := make([]byte, 0, len(r.body)+128)
	bytes = append(bytes, "HTTP/1.1 "...)
	bytes = strconv.AppendInt(bytes, int64(r.statusCode), 10)
	bytes = append(bytes, ' ')
	bytes = append(bytes, r.reason...)
	bytes = append(bytes, '\r', '\n')

	connectionHeaders := make([]Header, 0, 1)
	for _, header := range r.headers {
		if strings.EqualFold(header.Name, "Content-Length") {
			continue
		}
		if strings.EqualFold(header.Name, "Connection") {
			connectionHeaders = append(connectionHeaders, header)
			continue
		}
		bytes = appendHeader(bytes, header)
	}

	bytes = append(bytes, "Content-Length: "...)
	bytes = strconv.AppendInt(bytes, int64(len(r.body)), 10)
	bytes = append(bytes, '\r', '\n')
	for _, header := range connectionHeaders {
		bytes = appendHeader(bytes, header)
	}
	bytes = append(bytes, '\r', '\n')
	return append(bytes, r.body...)
}

func appendHeader(bytes []byte, header Header) []byte {
	bytes = append(bytes, header.Name...)
	bytes = append(bytes, ':', ' ')
	bytes = append(bytes, header.Value...)
	return append(bytes, '\r', '\n')
}

// Write serializes response and continues until all bytes have been written.
func Write(writer io.Writer, response Response) error {
	bytes := response.Bytes()
	for len(bytes) > 0 {
		count, err := writer.Write(bytes)
		if err != nil {
			return err
		}
		if count == 0 {
			return io.ErrShortWrite
		}
		bytes = bytes[count:]
	}
	return nil
}

File: rust/src/response.rs
/// A limited byte-oriented HTTP/1.1 response used by this course.
pub struct Response {
    status_code: u16,
    reason: &'static str,
    content_type: &'static str,
    body: Vec<u8>,
}

impl Response {
    /// Returns the course's fixed text response.
    pub fn basic_ok() -> Self {
        Self {
            status_code: 200,
            reason: "OK",
            content_type: "text/plain",
            body: b"Hello, World!".to_vec(),
        }
    }

    pub(crate) fn image_bmp_ok(body: Vec<u8>) -> Self {
        Self {
            status_code: 200,
            reason: "OK",
            content_type: "image/bmp",
            body,
        }
    }

    pub(crate) fn bad_request() -> Self {
        Self {
            status_code: 400,
            reason: "Bad Request",
            content_type: "text/plain",
            body: Vec::new(),
        }
    }

    /// Serializes a status line, headers, header terminator, and body.
    pub fn bytes(&self) -> Vec<u8> {
        let header = format!(
            "HTTP/1.1 {} {}\r\nContent-Type: {}\r\nContent-Length: {}\r\nConnection: close\r\n\r\n",
            self.status_code,
            self.reason,
            self.content_type,
            self.body.len()
        );
        let mut bytes = Vec::with_capacity(header.len() + self.body.len());
        bytes.extend_from_slice(header.as_bytes());
        bytes.extend_from_slice(&self.body);
        bytes
    }
}

#[cfg(test)]
mod tests {
    use super::Response;

    #[test]
    fn basic_ok_serializes_a_byte_exact_response() {
        assert_eq!(
            Response::basic_ok().bytes(),
            b"HTTP/1.1 200 OK\r\nContent-Type: text/plain\r\nContent-Length: 13\r\nConnection: close\r\n\r\nHello, World!"
        );
    }
}
